### PR TITLE
Correction makes millis() exact for several speeds

### PR DESCRIPTION
**Pull request by cburstedde**
PR originally created against MiniCore, now moved here.

### CODE REVIEW

**Comment by MCUdude** [comment]
Haven't had the chance to look through this PR yet. Quick question though, will 32 MHz work? Even though none of my cores supports this frequency, the timing functions do.

**Comment by cburstedde** [comment]
> Haven't had the chance to look through this PR yet. Quick question though, will 32 MHz work? Even though none of my cores supports this frequency, the timing functions do.

millis () for 32 MHz is unaffected, since it was exact before already.

**Comment by cburstedde** [comment]
Done with 22.1184 (untested). I did not update the delay functions, just millis and micros.

Turns out 18.432 is actually the most recommended of the odd ones with micros() accuracy of 125ppm.

**Comment by cburstedde** [comment]
So I think I'm done with millis. There was a non-obvious roundoff in one of the typical F_CPU scaling macros, which accounted for several percent of error for some non-power-of-two clocks. Fixed this and added fine tuning for exactly zero long-term drift.

Tested with three different F_CPU settings on my laptop, with 18.432 MHz below 100ppm error by manual timing.

**Comment by cburstedde** [comment]
The millis() correction changes just the ISR. Its number of cycles will likely not be constant for most odd frequencies, but as you said that was never the case. The additional cost of this PR is an unsigned char compare and increment, which I'd judge acceptable going from an error in the percent range to zero. But please check it out and let me know what you think.

I've only got two crystals now, but I've tested several of the other odd frequencies by re#defining F_CPU. Seems alright.

**Comment by MCUdude** [comment]
Here are some results when testing with a signal generator at various frequencies. The way I know how long the ISR lasted is by setting PORTD = 0xFF as the first thing in the ISR, and setting PORTD = 0x00 as the last thing. Tested on an Atmega88, since this was a board I had within reach. But any AVR would do.

Here is how many instructions each ISR takes with including setting and clearing PORTD:

| F_CPU       | Cycles in ISR (old implementation) | Cycles in ISR (new implementation) |
|-------------|------------------------------------|------------------------------------|
| 1 MHz       | Jumping between 49 and 54          | Jumping between 51 and 52          |
| 1.8432 MHz  | Jumping between 49 and 54          | Jumping between 55, 59 and 61      |
| 2 MHz       | Jumping between 49 and 54          | Jumping between 51 and 52          |
| 3.6864 MHz  | ...                                | Jumping between 62, 63, 65 and 68  |
| 4 MHz       | ...                                | Jumping between 51 and 52          |
| 7.3728 MHz  |                                    | Jumping between 62, 65, 66 and 68  |
| 8 MHz       |                                    | Jumping between 51 and 52          |
| 11.0592 MHz |                                    | Jumping between 62, 63, 65 and 68  |
| 12 MHz      |                                    | Jumping between 55, 57, 59 and 61  |
| 14.7456 MHz |                                    | Jumping between 55 and 61          |
| 16 MHz      |                                    | Jumping between 51 and 52          |
| 18.4320 MHz |                                    | Jumping between 55 and 59          |
| 20 MHz      |                                    | Jumping between 55, 59, 63 and 66  |

Bottom line, this looks very consistent. Actually, the 20 MHz ISR was the one with the most jitter.

But how did you set up a way to long-term test the accuracy of millis()? Did you just print the millis() value to the serial monitor every minute or so, or did you do something else? One problem with this approach is that when the ISR fires while it's waiting (in a delay() or something) the microcontroller doesn't subtract the amount of time it took to execute the ISR from the delay function, and over time, timing will be slightly inaccurate.

**Comment by cburstedde** [comment]
Hi, thanks for going through this measurement! Happy the formulas work out ok.

The 20 MHz formula is the weirdest, since it uses a counter that rolls on reaching 5 and an increment for odd counts (2 out of 5). I can ponder this one and see if I find something more elegant.

i did not use delay(), since I know that it misses cycles long-term. What I did is run a blink with a 3-second sequence based on millis() alone, and hit a time measurement command on my linux laptop like `date "+%c %s.%N"`. With a bit of practice, the hit error is about 1/10 of a second. Then I take the difference between some current time and the initial time and divide that by 3. The fractional part of the result must be close to 0 +/- say 0.0xy, or something is wrong. It takes several hours for this error to grow closer to 1, when the lock-on gets lost. While still locked, the relative error is about 1. - ((tnow - tbegin) / 3.) / (round ((tnow - tbegin) / 3.)), which approaches 100ppm after some dozen minutes.

This is the loop() code, with unsigned long bmillis = 1000UL, which does not suffer from interrupts:

        unsigned long mnow = millis ();
        unsigned long tdif = mnow - mlast;
        if (tdif >= bmillis) {
                /* run millis blinker */
                if (++mstate == 3) {
                        mstate = 0;
                }
                analogWrite (ledpin[0], ledvals[mstate]);
                mlast += bmillis;
        }

**Comment by MCUdude** [comment]
> The 20 MHz formula is the weirdest, since it uses a counter that rolls on reaching 5 and an increment for odd counts (2 out of 5). I can ponder this one and see if I find something more elegant.

Great! I'll keep this open a little longer then. FYI I will be leaving my scope and my signal generator for a few days around Christmas. I'm able to continue testing tomorrow evening and a little bit on Wednesday morning.

> i did not use delay(), since I know that it misses cycles long-term. What I did is run a blink with a 3-second sequence based on millis() alone, and hit a time measurement command on my linux laptop like date "+%c %s.%N". With a bit of practice, the hit error is about 1/10 of a second. Then I take the difference between some current time and the initial time and divide that by 3. The fractional part of the result must be close to 0 +/- say 0.0xy, or something is wrong. It takes several hours for this error to grow closer to 1, when the lock-on gets lost. While still locked, the relative error is about 1. - ((tnow - tbegin) / 3.) / (integral part of ((tnow - tbegin) / 3.)), which approaches 100ppm after some dozen minutes.

Wow, that sounds like a very comprehensive testing method! Since your "setup" is a little difficult for me to set up (and wrap my head around, to be honest), are you confident that the other untested frequencies are reasonably accurate? I mean, you have already tested with 18.432 MHz, which is one of those "weird" baud rate friendly clocks, so you have already proven that your math works 👍

**Comment by cburstedde** [comment]
I've pushed an alternate 20 MHz to cburstedde@631a3495d1d4a6821ae5c7c4d93f83aa5659a99a.
How does this one behave?

**Comment by cburstedde** [comment]
The 20 MHz drift is well within spec. I'd say the benefit of removing the general roundoff beats the risk of overlooking one of the fine-tuned corrections. While only owning two crystals, I've checked many other frequencies by re#defining F_CPU and rescaling the measurements.

**Comment by MCUdude** [comment]
> I've pushed an alternate 20 MHz to cburstedde/MCUdude_corefiles@631a349.
How does this one behave?

Well, it now jumps between 58, 59, 62, 63, and 66 cycles to execute the ISR. So it now has five different execution times, but the maximum difference between the slowest and fastest is now 8 cycles instead of 11.

**Comment by cburstedde** [comment]
> > I've pushed an alternate 20 MHz to [cburstedde/MCUdude_corefiles@631a349](https://github.com/cburstedde/MCUdude_corefiles/commit/631a349).
> > How does this one behave?
> 
> Well, it now jumps between 58, 59, 62, 63, and 66 cycles to execute the ISR. So it now has five different execution times, but the maximum difference between the slowest and fastest is now 8 cycles instead of 11.

Assuming these occurrences are equally frequent, this is slower by 1 cycle on average. Let's stick with the PR?
Btw. it's running right now below 100ppm accuracy.

**Comment by cburstedde** [comment]
I've just tested the 12 MHz again and it is fine, too. We have now tested all 4 of the correction variants that I have implemented, so I'm voting that the code is working.

**Comment by MCUdude** [comment]
> Assuming these occurrences are equally frequent, this is slower by 1 cycle on average. Let's stick with the PR?
Btw. it's running right now below 100ppm accuracy.

It turns out that my scope can measure the average pulse width, and therefore the number of cycles. On average, less than 0.1 cycles separates them. Let's stick with the implementation in this PR.

**Comment by MCUdude** [comment]
> We have now tested all 4 of the correction variants that I have implemented, so I'm voting that the code is working.

Indeed. If you give me a thumbs up on this, I'll merge these three PRs right away

**Comment by cburstedde** [comment]
> > We have now tested all 4 of the correction variants that I have implemented, so I'm voting that the code is working.
> 
> Indeed. If you give me a thumbs up on this, I'll merge these three PRs right away

Let's go!
